tools/unitctl/unit-client-rs/src/unit_client.rs

```diff
@@ -223,72 +211,73 @@ impl UnitClient {
-                    detail: error.get("detail").unwrap_or(&"".into()).to_string(),
-                });
-            }
-            serde_json::from_reader(reader).map_err(|error| UnitClientError::JsonError {
-                source: error,
+                })?;
+
+            return Err(UnitClientError::HttpResponseJsonBodyError {
+                status,
                 path: path.to_string(),
-            })
+                error: error.get("error").unwrap_or(&"Unknown error".into()).to_string(),
+                detail: error.get("detail").unwrap_or(&"".into()).to_string(),
+            });
+        }
+        serde_json::from_reader(reader).map_err(|error| UnitClientError::JsonError {
+            source: error,
+            path: path.to_string(),
         })
     }
 
     pub fn listeners_api(&self) -> Box<dyn ListenersApi + 'static> {
         new_openapi_client!(self, ListenersApiClient, ListenersApi)
     }
 
-    pub fn listeners(&self) -> Result<HashMap<String, ConfigListener>, Box<UnitClientError>> {
-        let list_listeners = self.listeners_api().get_listeners();
-        self.execute_openapi_future(list_listeners)
-    }
-
-    pub fn execute_openapi_future<F: Future<Output = Result<R, OpenAPIError>>, R: for<'de> serde::Deserialize<'de>>(
-        &self,
-        future: F,
-    ) -> Result<R, Box<UnitClientError>> {
-        self.rt.block_on(future).map_err(|error| {
-            let remapped_error = if let OpenAPIError::Hyper(hyper_error) = error {
-                UnitClientError::new(hyper_error, self.control_socket.to_string(), "".to_string())
+    pub async fn listeners(&self) -> Result<HashMap<String, ConfigListener>, Box<UnitClientError>> {
+        self.listeners_api().get_listeners().await.or_else(|err| {
+            if let OpenAPIError::Hyper(hyper_error) = err {
+                Err(Box::new(UnitClientError::new(
+                    hyper_error,
+                    self.control_socket.to_string(),
+                    "".to_string(),
+                )))
             } else {
-                UnitClientError::OpenAPIError { source: error }
-            };
-
-            Box::new(remapped_error)
+                Err(Box::new(UnitClientError::OpenAPIError { source: err }))
+            }
         })
     }
 
     pub fn status_api(&self) -> Box<dyn StatusApi + 'static> {
         new_openapi_client!(self, StatusApiClient, StatusApi)
     }
 
-    pub fn status(&self) -> Result<Status, Box<UnitClientError>> {
-        let status = self.status_api().get_status();
-        self.execute_openapi_future(status)
+    pub async fn status(&self) -> Result<Status, Box<UnitClientError>> {
+        self.status_api().get_status().await.or_else(|err| {
+            if let OpenAPIError::Hyper(hyper_error) = err {
+                Err(Box::new(UnitClientError::new(
+                    hyper_error,
+                    self.control_socket.to_string(),
+                    "".to_string(),
```

Comment:
Same as above, do we want to add the path, so that the error message is more useful?

Reply:
added